File: Maximum Number of Balloons - Leetcode 1189.py
class Solution:
    def maxNumberOfBalloons(self, text: str) -> int:
        counter = defaultdict(int)
        balloon = "balloon"
        
        for c in text:
            if c in balloon:
                counter[c] += 1
        
        if any(c not in counter for c in balloon):
            return 0
        else:
            return min(counter["b"], counter["a"], counter["l"] // 2, counter["o"] // 2, counter["n"])

# Time Complexity: O(n)
# Space Complexity: O(1)

File: Maximum Number of Balloons - Leetcode 1189/Maximum Number of Balloons - Leetcode 1189.cpp
#include <unordered_map>
#include <algorithm>
#include <string>
using namespace std;

class Solution {
public:
    int maxNumberOfBalloons(string text) {
        unordered_map<char, int> counter;
        string balloon = "balloon";

        for (char c : text) {
            if (balloon.find(c) != string::npos) {
                counter[c]++;
            }
        }

        if (counter.find('b') == counter.end() || 
            counter.find('a') == counter.end() || 
            counter.find('l') == counter.end() || 
            counter.find('o') == counter.end() || 
            counter.find('n') == counter.end()) {
            return 0;
        } else {
            return min({counter['b'], counter['a'], counter['l'] / 2, counter['o'] / 2, counter['n']});
        }
    }
};
